ydb/core/kqp/provider/rewrite_io_utils.cpp

```diff
@@ -0,0 +1,169 @@
+}
+
+void SaveQueryGraph(const TExprNode::TPtr& carrier, TExprContext& ctx, const TExprNode::TPtr& payload) {
+    AddChild(carrier, ctx.NewCallable(payload->Pos(), QueryGraphNodeSignature, {payload}));
+}
+
+void InsertExecutionOrderDependencies(const TExprNode::TPtr& queryGraph, const TExprNode::TPtr& worldBefore) {
+    VisitExpr(
+        queryGraph,
+        nullptr,
+        [&worldBefore](const TExprNode::TPtr& node) {
+            if (node->ChildrenSize() > 0 && node->Child(0)->IsWorld()) {
+                ChangeChild(node, 0, worldBefore);
+            }
+            return true;
+        }
+    );
+}
+
+bool CheckTopLevelness(const TExprNode::TPtr& candidateRead, const TExprNode::TPtr& queryGraph) {
+    THashSet<TExprNode::TPtr> readsInCandidateSubgraph;
+    VisitExpr(candidateRead, [&readsInCandidateSubgraph](const TExprNode::TPtr& node) {
+        if (node->IsCallable(ReadName)) {
+            readsInCandidateSubgraph.emplace(node);
+        }
+        return true;
+    });
+
+    return !FindNode(queryGraph, [&readsInCandidateSubgraph](const TExprNode::TPtr& node) {
+        return node->IsCallable(ReadName) && !readsInCandidateSubgraph.contains(node);
+    });
+}
+
+TExprNode::TPtr FindTopLevelRead(const TExprNode::TPtr& queryGraph) {
+    const TExprNode::TPtr* lastReadInTopologicalOrder = nullptr;
+    VisitExpr(
+        queryGraph,
+        nullptr,
+        [&lastReadInTopologicalOrder](const TExprNode::TPtr& node) {
+            if (node->IsCallable(ReadName)) {
+                lastReadInTopologicalOrder = &node;
+            }
+            return true;
+        }
+    );
+
+    if (!lastReadInTopologicalOrder) {
+        return nullptr;
+    }
+
+    YQL_ENSURE(CheckTopLevelness(*lastReadInTopologicalOrder, queryGraph),
+               "Info for developers: assumption that there is only one top level Read! is wrong "
+               "for the expression graph of the query stored in the view:\n"
+                   << queryGraph->Dump());
+
+    return *lastReadInTopologicalOrder;
+}
+
+}
+
+TExprNode::TPtr RewriteReadFromView(
+    const TExprNode::TPtr& node,
+    TExprContext& ctx,
+    const TString& query,
+    const TString& cluster
+) {
+    const auto readNode = node->ChildPtr(0);
+    const auto worldBeforeThisRead = readNode->ChildPtr(0);
+
+    TExprNode::TPtr queryGraph = FindSavedQueryGraph(readNode);
+    if (!queryGraph) {
+        queryGraph = CompileQuery(query, ctx, cluster);
+        if (!queryGraph || ContainsNullNode(queryGraph)) {
+            ctx.AddError(TIssue(readNode->Pos(ctx),
+                         "The query stored in the view contains errors and cannot be compiled."));
+            return nullptr;
+        }
+        YQL_CLOG(TRACE, ProviderKqp) << "Expression graph of the query stored in the view:\n"
+                                     << NCommon::ExprToPrettyString(ctx, *queryGraph);
+
+        InsertExecutionOrderDependencies(queryGraph, worldBeforeThisRead);
```

Comment:
This looks a bit like a hack. Parser should be able to return you all "root" world dependences, so you don't have to search for the in query graph.

Reply:
I don't know of a way to do it. I will ask Vitaliy Stoyan to review this change too